### PR TITLE
Refactors genetic meltdowns into datums

### LINKED ISSUE #83592: Dullahan race from Genetic instability is broken
## Issue Summary

While I was playing with genetic instability mutation I rolled a Dullahan race and it was bugged out. My body was not connected to the head. I could see with my body but could not hear anything or see any gesture. I hope this race gets a bit fixed (i know it was problematic before) or just cut out of genetic instabitily pool. You can continue rolling races and in a normal round without lavaland mirror you are stuck with a body that can interact with other players but that cant see or hear the reply back.

## Round ID:
[229714](https://statbus.space/rounds/229714)

### PR DESCRIPTION
**Pull request by SmArtKar**
## About The Pull Request

Genetic meltdowns now use datums instead of a prob and switch monstrocity, making them far more readable. Requested by Jacquerel on discord.

Additionally, removed 0.83% chance for a meltdown to turn you into a dullahan, as those are extremely broken right now.
Closes #83592

## Why It's Good For The Game

This system is much nicer, readable and more extendable than the current one.

## Changelog
:cl:
del: You can no longer get turned into a dullahan by a genetic meltdown
refactor: Refactored genetic meltdown code into datums, making it more extendable.
/:cl:

### CODE REVIEW

**Comment by EuSouAFazer** [comment]
thank god fuck dullahams

**Comment by moocowswag** [comment]
Broken as in doesnt work/not maintained or unbalanced, cause it seems in line with arcade pulse pistol, rare maint loot, die of fate, etc

File: code/datums/instability_meltdown.dm [hunk 1]
```diff
@@ -0,0 +1,175 @@
+/datum/instability_meltdown
+	/// How likely a meltdown is to be picked
+	var/meltdown_weight = 0
```
Comment (Jacquerel) [inline code/datums/instability_meltdown.dm:3]: If almost all of these are `1` why didn't you default it to 1?

File: code/datums/instability_meltdown.dm [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	meltdown_weight = 1
+
+/datum/instability_meltdown/organ_vomit/meltdown(mob/living/carbon/human/victim)
+	var/list/elligible_organs = list()
+	for(var/obj/item/organ/internal/internal_organ in victim.organs) //make sure we dont get an implant or cavity item
+		elligible_organs += internal_organ
+	victim.vomit(VOMIT_CATEGORY_DEFAULT, lost_nutrition = 10)
+	if(!elligible_organs.len)
+		return
+	var/obj/item/organ/O = pick(elligible_organs)
```
Comment (Jacquerel) [inline code/datums/instability_meltdown.dm:85]: While you're touching this give it a better var name

File: code/datums/instability_meltdown.dm (hunk 1 above)
Comment (Jacquerel) [reply code/datums/instability_meltdown.dm:3]: I am not 100% sure what adding something with a weight of 0 does but I don't think it actually makes it _totally_ impossible for it to roll, unintuitively.
If this is to support the abstract fatal type, you might need to add an abstractness check as well (`var/abstract_type = typepath` then check against that path is a pattern we often use, so it doesn't need to be overridden in all children)

**Review by Jacquerel — CHANGES_REQUESTED** [review]

**Comment by Jacquerel** [comment]
> Broken as in doesnt work/not maintained or unbalanced, cause it seems in line with arcade pulse pistol, rare maint loot, die of fate, etc
> 
> Nevermind i read the linked issue, im too used to the league of legends context of “broken”

Dullahans have only been barely functional most halloweens, often nonfunctional, and get worse every time.

File: code/datums/dna.dm [hunk 3]
```diff
@@ -871,81 +872,17 @@ GLOBAL_LIST_INIT(total_uf_len_by_block, populate_total_uf_len_by_block())
...
-				addtimer(CALLBACK(src, PROC_REF(something_horrible_mindmelt)), 3 SECONDS)
-			if(6)
-				slow_psykerize()
+
+	var/list/valid_meltdowns = list()
+	var/nonfatal = prob(max(70-instability,0))
+
+	for (var/datum/instability_meltdown/meltdown_type as anything in dna.meltdown_cache)
+		if (initial(meltdown_type.fatal) != nonfatal && initial(meltdown_type.abstract_type) != meltdown_type)
+			valid_meltdowns[meltdown_type] = initial(meltdown_type.meltdown_weight)
```
Comment (Time-Green) [inline code/datums/dna.dm:881]: tbh if you're caching all this you might as well cache this as well. Keep the possible types as a static list, and only generate it the first time by checking the length

File: code/datums/instability_meltdown.dm [hunk 4]
```diff
@@ -0,0 +1,161 @@
+/datum/instability_meltdown
```
Comment (Time-Green) [inline code/datums/instability_meltdown.dm:1]: Autodoc the type please

File: code/datums/instability_meltdown.dm [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+	victim.investigate_log("has been transformed into a statue by DNA instability.", INVESTIGATE_DEATHS)
+	victim.death()
+	victim.petrify(statue_timer = INFINITY, save_brain = FALSE)
+	victim.ghostize(FALSE)
+
+/// Either dismembers you, or if unable to, gibs you
+/datum/instability_meltdown/fatal/dismember
+
+/datum/instability_meltdown/fatal/dismember/meltdown(mob/living/carbon/human/victim)
+	var/obj/item/bodypart/BP = victim.get_bodypart(pick(BODY_ZONE_CHEST,BODY_ZONE_HEAD))
```
Comment (Time-Green) [inline code/datums/instability_meltdown.dm:133]:
```suggestion
	var/obj/item/bodypart/part= victim.get_bodypart(pick(BODY_ZONE_CHEST, BODY_ZONE_HEAD))
```
write the varnames out ('part' or something is fine)

File: code/datums/instability_meltdown.dm [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+/datum/instability_meltdown/alright/meltdown(mob/living/carbon/human/victim)
+	to_chat(victim, span_notice("Oh, I actually feel quite alright!"))
+
+/// Gives you the same text as above but now when you're hit you take 200 times more damage
+/datum/instability_meltdown/not_alright
+
+/datum/instability_meltdown/not_alright/meltdown(mob/living/carbon/human/victim)
+	to_chat(victim, span_notice("Oh, I actually feel quite alright!"))
+	victim.physiology.damage_resistance -= 20000 //you thought
```
Comment (Time-Green) [inline code/datums/instability_meltdown.dm:46]: can you log this btw? its pretty severe

**Review by Time-Green — CHANGES_REQUESTED** [review]
Very nice! Minor suggestions

File: code/datums/dna.dm [hunk 7]
```diff
@@ -79,7 +79,12 @@ GLOBAL_LIST_INIT(total_uf_len_by_block, populate_total_uf_len_by_block())
 	var/stability = 100
 	///Did we take something like mutagen? In that case we cant get our genes scanned to instantly cheese all the powers.
 	var/scrambled = FALSE
-
+	/// List of potential genetic meltdowns
+	var/static/list/meltdown_cache = typecacheof(/datum/instability_meltdown, ignore_root_path = TRUE)
```
Comment (Time-Green) [inline code/datums/dna.dm:83]: you can drop this cache now btw, its only used once

File: code/datums/dna.dm [hunk 8]
```diff
@@ -871,81 +874,23 @@ GLOBAL_LIST_INIT(total_uf_len_by_block, populate_total_uf_len_by_block())
...
-					addtimer(CALLBACK(src, PROC_REF(death)), 3 SECONDS)
-			if(5)
-				to_chat(src, span_phobia("LOOK UP!"))
-				addtimer(CALLBACK(src, PROC_REF(something_horrible_mindmelt)), 3 SECONDS)
-			if(6)
-				slow_psykerize()
+
+	var/nonfatal = prob(max(70-instability, 0))
+
+	if(!LAZYLEN(dna.nonfatal_meltdowns))
```
Comment (Time-Green) [inline code/datums/dna.dm:880]:
```suggestion
	if(!dna.nonfatal_meltdowns.len)
```
Lazylen is only if you don't know if there's a list instance, but you instantiate it as a list() so .len is 'better' here

**Pull request merged**